File: Dev/unitTest_Engine_cs/ObjectSystem2D/AddComponentByComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ace;

namespace unitTest_Engine_cs.ObjectSystem2D
{
	class AddComponentByComponent : EngineTest
	{
		class ComponentAddingObject : Object2DComponent
		{
			int count = 0;
			protected override void OnUpdate()
			{
				count++;
				if(count == 10)
				{
					Owner.AddComponent(new ComponentAddingObject(), "Add");
					Console.WriteLine("vanish object component.");
					Vanish();
				}
			}
		}

		class ComponentAddingLayer : Layer2DComponent
		{
			int count = 0;
			protected override void OnUpdated()
			{
				count++;
				if(count == 10)
				{
					Owner.AddComponent(new ComponentAddingLayer(), "Add");
					Console.WriteLine("vanish layer component.");
					Vanish();
				}
			}
		}

		class ComponentAddingScene : SceneComponent
		{
			int count = 0;
			protected override void OnUpdated()
			{
				count++;
				if(count == 10)
				{
					Owner.AddComponent(new ComponentAddingScene(), "Add");
					Console.WriteLine("vanish scene component.");
					Vanish();
				}
			}
		}

		class ObjectAddingObject : TextureObject2D
		{
			int count = 0;
			protected override void OnUpdate()
			{
				count++;
				if(count == 10)
				{
					Layer.AddObject(new ObjectAddingObject());
					Console.WriteLine("vanish object.");
					Vanish();
				}
			}
		}

		public AddComponentByComponent()
			: base(60)
		{

		}

		protected override void OnStart()
		{
			var scene = new Scene();
			var layer = new Layer2D();
			var obj = new ObjectAddingObject();

			Engine.ChangeScene(scene);
			scene.AddLayer(layer);
			layer.AddObject(obj);

			scene.AddComponent(new ComponentAddingScene(), "Add");
			layer.AddComponent(new ComponentAddingLayer(), "Add");
			obj.AddComponent(new ComponentAddingObject(), "Add");
		}
	}
}

File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/AddComponentByComponent.cpp
#include <ace.h>
#include <gtest/gtest.h>
#include <memory>
#include "../EngineTest.h"

using namespace std;
using namespace ace;

class ObjectSystem_AddComponentByComponent : public EngineTest
{
	class AddingObjectComponent : public Object2DComponent
	{
		int count = 0;
	protected:
		void OnUpdate() override
		{
			count++;
			if (count == 10)
			{
				GetOwner()->AddComponent(make_shared<AddingObjectComponent>(), ToAString("Add"));
				printf("vanish object component.\n");
				Vanish();
			}
		}
	};

	class AddingLayerComponent : public Layer2DComponent
	{
		int count = 0;
	protected:
		void OnUpdate() override
		{
			count++;
			if (count == 10)
			{
				GetOwner()->AddComponent(make_shared<AddingLayerComponent>(), ToAString("Add"));
				Vanish();
			}
		}
	};

	class AddingSceneComponent : public SceneComponent
	{
		int count = 0;
	protected:
		void OnUpdate() override
		{
			count++;
			if (count == 10)
			{
				GetOwner()->AddComponent(make_shared<AddingSceneComponent>(), ToAString("Add"));
				Vanish();
			}
		}
	};

public:
	ObjectSystem_AddComponentByComponent(bool isOpenGLMode)
		: EngineTest(ace::ToAString("AddComponentByComponent"), isOpenGLMode, 60)
	{
	}

protected:
	void OnStart()
	{
		auto scene = std::make_shared<ace::Scene>();
		auto layer = std::make_shared<ace::Layer2D>();
		auto obj = std::make_shared<ace::TextureObject2D>();

		ace::Engine::ChangeScene(scene);
		scene->AddLayer(layer);
		layer->AddObject(obj);

		//scene->AddComponent(make_shared<AddingSceneComponent>(), ToAString("Add"));
		//layer->AddComponent(make_shared<AddingLayerComponent>(), ToAString("Add"));
		obj->AddComponent(make_shared<AddingObjectComponent>(), ToAString("Add"));
	}
};

ENGINE_TEST(ObjectSystem, AddComponentByComponent)